File: examples3d/spring_joints3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Fixed ground to attach one end of the joints.
     */
    let rigid_body = RigidBodyBuilder::fixed();
    let ground_handle = world.insert_body(rigid_body);

    /*
     * Spring joints with a variety of spring parameters.
     * The middle one has uses critical damping.
     */
    let num = 30;
    let radius = 0.5;
    let mass = Ball::new(radius).mass_properties(1.0).mass();
    let stiffness = 1.0e3;
    let critical_damping = 2.0 * (stiffness * mass).sqrt();
    for i in 0..=num {
        let x_pos = -6.0 + 1.5 * i as f32;
        let ball_pos = Vector::new(x_pos, 4.5, 0.0);
        let rigid_body = RigidBodyBuilder::dynamic()
            .translation(ball_pos)
            .can_sleep(false);
        let collider = ColliderBuilder::ball(radius);
        let (handle, _) = world.insert(rigid_body, collider);

        let damping_ratio = i as f32 / (num as f32 / 2.0);
        let damping = damping_ratio * critical_damping;
        let joint = SpringJointBuilder::new(0.0, stiffness, damping)
            .local_anchor1(ball_pos - Vector::Y * 3.0);
        world.insert_impulse_joint(ground_handle, handle, joint);

        // Box that will fall on to of the springed balls, makes the simulation funnier to watch.
        let rigid_body = RigidBodyBuilder::dynamic().translation(ball_pos + Vector::Y * 5.0);
        let collider = ColliderBuilder::cuboid(radius, radius, radius).density(100.0);
        world.insert(rigid_body, collider);
    }

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec3::new(15.0, 5.0, 42.0), Vec3::new(13.0, 1.0, 1.0));
}

File: python/rapier-testbed/rapier_testbed/examples3/spring_joints3.py
"""Port of examples3d/spring_joints3.rs."""
from __future__ import annotations

import math

import rapier3d as rp
from .._registry import register

CATEGORY = "Joints"
NAME = "Spring Joints"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_h = bodies.insert(rp.RigidBody.fixed())

    num = 30
    radius = 0.5
    # Ball mass with default density (1.0) is (4/3 pi r^3).
    mass = rp.Ball(radius).__class__  # not strictly used, but mirror the original
    # Mirror the Rust formula: mass = Ball::new(r).mass_properties(1.0).mass()
    mass = rp.SharedShape.ball(radius).mass_properties(1.0).mass
    stiffness = 1.0e3
    critical_damping = 2.0 * math.sqrt(stiffness * mass)

    for i in range(num + 1):
        x_pos = -6.0 + 1.5 * i
        ball_pos = (x_pos, 4.5, 0.0)
        body = rp.RigidBody.dynamic().translation(ball_pos).can_sleep(False)
        h = bodies.insert(body)
        colliders.insert_with_parent(rp.Collider.ball(radius), h, bodies)

        damping_ratio = i / (num / 2.0)
        damping = damping_ratio * critical_damping
        joint = (
            rp.SpringJoint.builder(0.0, stiffness, damping)
            .local_anchor1((ball_pos[0], ball_pos[1] - 3.0, ball_pos[2]))
            .build()
        )
        impulse_joints.insert(ground_h, h, joint, wake_up=True)

        box = rp.RigidBody.dynamic().translation((ball_pos[0], ball_pos[1] + 5.0, ball_pos[2]))
        bh = bodies.insert(box)
        colliders.insert_with_parent(
            rp.Collider.cuboid(radius, radius, radius).density(100.0), bh, bodies
        )

    testbed.set_world_with_params(
        bodies, colliders, impulse_joints, multibody_joints, (0.0, -9.81, 0.0)
    )
    testbed.look_at((15.0, 5.0, 42.0), (13.0, 1.0, 1.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
